src/RequestParser.cpp

```diff
@@ -788,6 +608,294 @@ void RequestParser::validateTransferEncoding(HTTPRequest& request)
+	if (request.hasBody && !isMethodAllowedToHaveBody(request)) {
+		request.httpStatus = StatusMethodNotAllowed;
+		throw std::runtime_error(ERR_UNEXPECTED_BODY);
+	}
+}
+
+/**
+ * @brief Validates the "Host" header in the given HTTP request.
+ *
+ * This function checks if the "Host" header is present and non-empty in the HTTP request.
+ * It also validates the format of the host value, ensuring it is either a valid IP address
+ * (with or without a port) or a valid hostname.
+ *
+ * @param request The HTTP request object containing headers to be validated.
+ *
+ * @throws std::runtime_error If the "Host" header is missing, empty, or contains an invalid value.
+ * The specific error message thrown depends on the type of validation failure:
+ * - ERR_MISSING_HOST_HEADER: The "Host" header is missing.
+ * - ERR_EMPTY_HOST_VALUE: The "Host" header value is empty.
+ * - ERR_INVALID_IP_WITH_PORT: The "Host" header contains an invalid IP address with a port.
+ * - ERR_INVALID_IP: The "Host" header contains an invalid IP address.
+ * - ERR_INVALID_HOSTNAME: The "Host" header contains an invalid hostname.
+ */
+void RequestParser::validateHostHeader(HTTPRequest& request)
+{
+	std::map<std::string, std::string>::const_iterator iter = request.headers.find("Host");
+	if (iter == request.headers.end()) {
+		request.httpStatus = StatusBadRequest;
+		throw std::runtime_error(ERR_MISSING_HOST_HEADER);
+	}
+
+	if (iter->second.empty()) {
+		request.httpStatus = StatusBadRequest;
+		throw std::runtime_error(ERR_EMPTY_HOST_VALUE);
+	}
+
+	bool hasPort = false;
+	if (isIPAddress(iter->second, hasPort)) {
+		if (hasPort) {
+			if (!webutils::isIpAddressValid(iter->second.substr(0, iter->second.find(':'))) ||
+			!webutils::isPortValid(iter->second.substr(iter->second.find(':') + 1))) {
+				request.httpStatus = StatusBadRequest;
+				throw std::runtime_error(ERR_INVALID_IP_WITH_PORT);
+			}
+		} else {
+			if (!webutils::isIpAddressValid(iter->second)) {
+				request.httpStatus = StatusBadRequest;
+				throw std::runtime_error(ERR_INVALID_IP);
+			}
+		}
+	} else {
+		if (!isValidHostname(iter->second)) {
+			request.httpStatus = StatusBadRequest;
+			throw std::runtime_error(ERR_INVALID_HOSTNAME);
+		}
+	}
+}
+
+/**
+ * @brief Validates that there are no multiple "Host" headers in the HTTP request.
+ *
+ * This function checks if the provided header name is "Host". If it is, it then checks
+ * if the "Host" header already exists in the request headers. If a "Host" header is
+ * found, it sets the HTTP status of the request to BadRequest and throws a runtime error.
+ *
+ * @param headerName The name of the header to validate.
+ * @param request The HTTP request object containing the headers.
+ *
+ * @throws std::runtime_error if multiple "Host" headers are found.
+ */
+void RequestParser::validateNoMultipleHostHeaders(const std::string& headerName, HTTPRequest& request)
+{
+	if (headerName == "Host") {
+		if (request.headers.find("Host") != request.headers.end()) {
```

Comment:
Field names are case-insensitive per RFC 9110 - 5.1 [Field Names](https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1-3)

This means if a client sends two headers with "Host = [...]" and "HOST = [...]" it wouldn't be caught.

General: I think we should sanitize the field names, e.g. all lowercase

Reply:
Good finding. Then I suggest while parsing the headers we send all field names through tolower() - or more precisely all of the letters - and then do the other checks.

Reply:
Added function in utilities to lowercase the parsed field name.

Reply:
Maybe we should also track this as issue for existing stuff?

I found in handleCompleteRequestHeader() a search for "Host", maybe there are more

Reply:
Changed to "host" in handleCompleteRequestHeader()